### PR TITLE
Add case specific map and traverse

### PR DESCRIPTION
**Pull request by Fresheyeball**

### CODE REVIEW

File: Data/These.hs [hunk 1]
```diff
@@ -174,6 +179,24 @@ mapThese f _ (This  a  ) = This (f a)
 mapThese _ g (That    x) = That (g x)
 mapThese f g (These a x) = These (f a) (g x)
 
+-- | Case specific map.
+mapThese' :: (a -> c) -> (b -> d) -> (a -> b -> (c,d)) -> These a b -> These c d
```
Comment (phadej) [inline Data/These.hs]: I'd need a motivating example

File: Data/These.hs [hunk 2]
```diff
@@ -250,9 +273,7 @@ instance Bifoldable1 These where
     bifold1 = these id id (<>)
 
 instance Bitraversable These where
-    bitraverse f _ (This x) = This <$> f x
-    bitraverse _ g (That x) = That <$> g x
-    bitraverse f g (These x y) = These <$> f x <*> g y
+    bitraverse = mapTheseM
```
Comment (phadej) [inline Data/These.hs]: :+1:

File: Data/These.hs [hunk 3]
```diff
@@ -174,6 +179,24 @@ mapThese f _ (This  a  ) = This (f a)
...
 mapThese f g (These a x) = These (f a) (g x)
 
+-- | Case specific map.
+mapThese' :: (a -> c) -> (b -> d) -> (a -> b -> (c,d)) -> These a b -> These c d
+mapThese' f _ _ (This  a  ) = This (f a)
+mapThese' _ g _ (That    b) = That (g b)
+mapThese' _ _ h (These a b) = uncurry These $ h a b
+
+-- | 'Bitraversable` map.
+mapTheseM :: Applicative f => (a -> f c) -> (b -> f d) -> These a b -> f (These c d)
```
Comment (phadej) [inline Data/These.hs]: `mapTheseA`. There are no `Monad`.

File: Data/These.hs (hunk 1 above)
Comment (Fresheyeball) [reply Data/These.hs]: I need to produce a value from `These` with different  logic for the `These` case. I have written this function a few times, and felt I needed to contribute it back.

File: Data/These.hs (hunk 3 above)
Comment (Fresheyeball) [reply Data/These.hs]: Ok, will change the name.

File: Data/These.hs [hunk 4]
```diff
@@ -174,6 +179,24 @@ mapThese f _ (This  a  ) = This (f a)
...
+mapThese' _ _ h (These a b) = uncurry These $ h a b
+
+-- | 'Bitraversable` map.
+mapTheseA :: Applicative f => (a -> f c) -> (b -> f d) -> These a b -> f (These c d)
+mapTheseA f _ (This  a  ) = This <$> f a
+mapTheseA _ g (That    b) = That <$> g b
+mapTheseA f g (These a b) = These <$> f a <*> g b
+
+-- | Case specific traverse.
+mapTheseF' :: Functor f => (a -> f c) -> (b -> f d) -> (a -> b -> f (c,d)) -> These a b -> f (These c d)
```
Comment (Fresheyeball) [inline Data/These.hs]: Small correction, I use this one more often than `mapThese'`.

**Comment by phadej** [comment]
I added `bitraverseThese` in https://hackage.haskell.org/package/these-0.7.5, still not sure if I really want to have other functions

**Comment by phadej** [comment]
I'd like to see better justifications (and naming pattern).

I added some combinators in `Data.These.Combinators` (see http://oleg.fi/haddocks/these/, #102)

**Pull request closed without merging**